### PR TITLE
Issue 1436: Indentation error on Jinja templated test case

### LINKED ISSUE #1436: Indentation error on Jinja templated test case
As discussed in #1430 the test file [`test/fixtures/linter/autofix/ansi/016_index_error_with_jinja_if/after.sql`](https://github.com/sqlfluff/sqlfluff/blob/62687a79887205a0c877ccdd8bf18be44e39738e/test/fixtures/linter/autofix/ansi/016_index_error_with_jinja_if/after.sql) creates an unexpected dedent when fixed with rule L003

## Expected Behaviour

The `{{ product }}` line is not dedented, and in fact should be indented

## Observed Behaviour

The `{{ product }}` line is dedented

## Steps to Reproduce

Run this:
```
sqlfluff fix test/fixtures/linter/autofix/ansi/016_index_error_with_jinja_if/before.sql --rules L003 -f
```

Do a `git diff` and note this output:

```
 {% for product in products %}
 SELECT
-  brand,
-  country_code,
-  category,
-  name,
-  id
+    brand,
+    country_code,
+    category,
+    name,
+    id
 FROM
-  {{ product }}
+{{ product }}
```

The select clauses have been correct indented, but the from clause has been dedented instead of indented.

Not this does NOT happen if `product` is used instead of `{{ product }}` and also does not happen if `{% if not loop.last %} UNION ALL {% endif %}` is used instead of `{% if not loop.last -%} UNION ALL {%- endif %}`

## Dialect

ansi with jinja

## Version
Include the output of `sqlfluff --version` along with your Python version

SQLFluff version: `main` as of September 23 (commit hash b977d63a8552a7d60e9310ef5efceeedb2123361) (to be updated after #1430 is merged
Python version: 3.8

## Configuration

No config so all defaults.

### PR DESCRIPTION
**Pull request by barrywhart**
### Brief summary of the change made

Fixes #1436 

Fixes and some refactoring of L003 to allow it to fix the indentation of templated code (Jinja, dbt), e.g. templated table name or `for` or `if`. Also fixes bugs in a few other areas of the code that were discovered during this work.

### Are there any other side effects of this change that we should be aware of?

This PR touches the indentation logic, one of the more visible areas of SQLFluff. It fixes some existing issues and limitations with automatic indentation. This code is quite complex. Pre-existing test coverage was good, but not 100%, so some new issues could surface after release. I feel fairly confident about this PR because the `test/rules/std_fix_auto_test.py` tests are pretty complex -- many are based on complex real-life templated code that drove the early development of template and indentation in SQLFluff.

### Pull Request checklist
- [X] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/parser` (note YML files can be auto generated with `python test/generate_parse_fixture_yml.py` or by running `tox` locally).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/core/templaters/jinja.py [hunk 1]
```diff
@@ -303,8 +303,8 @@ def process(
             )
             return None, violations
 
-    re_open_tag = re.compile(r"^{%[\+\-]?\s*")
-    re_close_tag = re.compile(r"\s*[\+\-]?%}$")
+    re_open_tag = re.compile(r"^\s*{%[\+\-]?\s*")
```
Comment (barrywhart) [inline src/sqlfluff/core/templaters/jinja.py:306]: Unrelated bug fix turned up by a dbt test: Sometimes the Jinja tags are preceded or followed by whitespace.

File: src/sqlfluff/rules/L003.py [hunk 2]
```diff
@@ -127,7 +127,7 @@ def _process_raw_stack(cls, raw_stack, tab_space_size=4):
                         # an increase in indentation? Can't quite
                         # remember the logic here. Let's go with that.
                         clean_indent = True
-                else:
+                elif not elem.is_type("placeholder"):
```
Comment (barrywhart) [inline src/sqlfluff/rules/L003.py:130]: L003 computes the desired indentation based on the first code on a line. Need to skip placeholders (template code) so it finds the first **SQL** code on the line.

File: src/sqlfluff/rules/L003.py [hunk 3]
```diff
@@ -255,9 +255,12 @@ def _eval(self, segment, raw_stack, memory, **kwargs):
             elif segment.segments or (segment.is_meta and segment.indent_val != 0):
                 # it's not a raw segment or placeholder. Carry on.
                 return LintResult(memory=memory)
+            elif segment.is_type("placeholder"):
```
Comment (barrywhart) [inline src/sqlfluff/rules/L003.py:258]: Similar fix as above.

File: test/fixtures/dbt/dbt_project/models/my_new_project/incremental.sql [hunk 4]
```diff
@@ -9,7 +9,7 @@ select
 from products
 inner join dispensaries
 where not products._fivetran_deleted
-{% if true -%}
-and products.valid_date_local >= (
-    select max(valid_date_local) from {{ this }})
-{% endif -%}
+    {% if true -%}
```
Comment (barrywhart) [inline test/fixtures/dbt/dbt_project/models/my_new_project/incremental.sql:12]: Hurray, a test "broke" with this change, but the new behavior looks better.

File: src/sqlfluff/rules/L003.py (hunk 3 above)
Comment (tunetheweb) [reply src/sqlfluff/rules/L003.py:258]: Any reason not to merge this into the above code? Especially as it specifically mentions placeholders?

Comment (barrywhart) [reply src/sqlfluff/rules/L003.py:258]: I wondered about that. The above comment looks potentially wrong to me. The comment mentions placeholders, but it doesn't seem to do anything placeholder related.

I'm going to have to rework this code before merge, because of some tests it broke. Let's revisit at that time.

I think rather than skip placeholders, we need to process them, but only if there's no regular code on that line.

File: src/sqlfluff/core/templaters/base.py [hunk 5]
```diff
@@ -302,7 +302,10 @@ def templated_slice_to_source_slice(
                 return slice(insertion_point, insertion_point)
             # It's within a segment.
             else:
-                if ts_start_subsliced_file[0][0] == "literal":
+                if (
+                    ts_start_subsliced_file
```
Comment (barrywhart) [inline src/sqlfluff/core/templaters/base.py:306]: During work on this PR, at some point it was crashing here with an `IndexError` because `ts_start_subsliced_file` was empty. Without really understanding what was going on, I tried the simplest possible fix -- checking first that the array is not empty. Seemed to work.

File: src/sqlfluff/core/templaters/jinja.py [hunk 6]
```diff
@@ -354,15 +354,20 @@ def _slice_template(cls, in_str: str) -> Iterator[RawFileSlice]:
...
-                    else:
+                    elif first_word in ("for", "if"):
                         block_type = "block_start"
-                        if trimmed_content.split()[0] == "for":
+                        if first_word == "for":
                             block_subtype = "loop"
+                    else:
+                        # Not necessarily a comment, but we can treat it as one
+                        # for our purposes.
+                        block_type = "comment"
```
Comment (barrywhart) [inline src/sqlfluff/core/templaters/jinja.py:370]: This change fixes an indentation issue -- a line like this:
```
{% set considered_actions=['a1', 'a2', 'a3', 'a4', 'a5'] %}
```

was previously labeled as `block_type` = `block_start`. That caused other parts of the code to be over-indented.

File: src/sqlfluff/rules/L003.py [hunk 7]
```diff
@@ -45,7 +45,7 @@ class Rule_L003(BaseRule):
     """
 
     _works_on_unparsable = False
-    config_keywords = ["tab_space_size", "indent_unit", "lint_templated_tokens"]
```
Comment (barrywhart) [inline src/sqlfluff/rules/L003.py:48]: I removed this setting and the related code because I found that removing it fixed one of my new tests. There are no existing tests for this feature and (apparently) no documentation other than its appearance in a sample config file, so I thought _maybe_ it was reasonable to remove it?

File: src/sqlfluff/rules/L003.py [hunk 8]
```diff
@@ -244,28 +348,23 @@ def _eval(self, segment, raw_stack, memory, **kwargs):
                 "comment_lines": [],
             }
 
-        if segment.is_type("newline"):
```
Comment (barrywhart) [inline src/sqlfluff/rules/L003.py:247]: I moved this section of code (with some changes) to a new function `_find_trigger()`.

File: src/sqlfluff/rules/L003.py [hunk 9]
```diff
@@ -244,28 +349,30 @@ def _eval(self, segment, raw_stack, memory, **kwargs):
...
         res = self._process_raw_stack(
-            raw_stack + (segment,), tab_space_size=self.tab_space_size
+            raw_stack, tab_space_size=self.tab_space_size, templated_file=templated_file
         )
+        if res:
+            return self._process_current_line(res, memory, templated_file)
+        else:
+            return LintResult(memory=memory)
+
+    def _process_current_line(self, res, memory, templated_file):
```
Comment (barrywhart) [inline src/sqlfluff/rules/L003.py:368]: I split this function out of the original `_eval()` function to make the code a little easier to understand and to separate concerns a bit. The main `_eval()` function processes each segment as it's passed in. OTOH, this function is only executed at the end of each **line** of code, to check the indentation.

File: src/sqlfluff/rules/L003.py [hunk 10]
```diff
@@ -328,27 +438,14 @@ def _eval(self, segment, raw_stack, memory, **kwargs):
...
-                    anchor=segment,
+                    anchor=trigger_segment,
                     memory=memory,
                     description="First line has unexpected indent",
                     fixes=[
                         LintFix("delete", elem) for elem in this_line["indent_buffer"]
                     ],
                 )
 
-        # Are we linting a placeholder, and if so are we allowed to?
```
Comment (barrywhart) [inline src/sqlfluff/rules/L003.py:339]: As noted earlier, removing this undocumented, untested, apparently not working code.

File: test/fixtures/linter/autofix/bigquery/003_templating/after.sql [hunk 11]
```diff
@@ -15,15 +15,15 @@ SELECT
         , campaign_count_{{action}}
     {% endfor %}
 FROM
-{% for action in considered_actions %}
+    {% for action in considered_actions %}
```
Comment (barrywhart) [inline test/fixtures/linter/autofix/bigquery/003_templating/after.sql:18]: I think this looks better.

File: src/sqlfluff/testing/rules.py [hunk 12]
```diff
@@ -62,7 +62,9 @@ def assert_rule_fail_in_sql(code, sql, configs=None, line_numbers=None):
     for e in lerrs:
         if e.desc().startswith("Unexpected exception"):
             pytest.fail(f"Linter failed with {e.desc()}")  # pragma: no cover
-    parse_errors = list(filter(lambda v: type(v) == SQLParseError, lerrs))
+    parse_errors = list(
```
Comment (barrywhart) [inline src/sqlfluff/testing/rules.py:65]: Fixes a bug where tests failed with poor error reporting on SQLTemplaterError

File: src/sqlfluff/testing/rules.py [hunk 13]
```diff
@@ -79,7 +81,11 @@ def assert_rule_fail_in_sql(code, sql, configs=None, line_numbers=None):
                 )
             )
     # The query should already have been fixed if possible so just return the raw.
-    return linted.tree.raw
+    if linted.num_violations(fixable=True) > 0:
+        fixed, _ = linted.fix_string()
```
Comment (barrywhart) [inline src/sqlfluff/testing/rules.py:85]: Fixes a bug where YAML rule tests (e.g. L003.yml) were not translating fixed code back to the templated version.

File: src/sqlfluff/rules/L003.py [hunk 14]
```diff
@@ -242,30 +338,66 @@ def _eval(self, segment, raw_stack, memory, **kwargs):
...
+            trigger_segment = memory["trigger"]
+            if trigger_segment:
+                # Not empty. Process it.
+                result = self._process_current_line(res, memory)
+                if segment.is_type("newline"):
+                    memory["trigger"] = None
+                return result
+        return LintResult(memory=memory)
+
+    def _process_current_line(self, res, memory):
```
Comment (barrywhart) [inline src/sqlfluff/rules/L003.py:393]: I split part of the `_eval()` function into this new function to make this rule easier to understand. Not technically necessary.

**Review by tunetheweb — APPROVED** [review]
I can't say I follow all the logic to be perfectly honest, but I can't spot anything wrong and am happy with the test case changes. So happy for this to be merged.
